### PR TITLE
azidentity multitenant authentication API

### PR DESCRIPTION
**Pull request by chlowell**
Usage looks like this:
```go
// cred will request tokens from the configured tenant unless another is specified in GetToken args
cred, err := azidentity.NewClientSecretCredential("primary-tenant", "client-id", "secret", &azidentity.ClientSecretCredentialOptions{
	// cred will (attempt to) authenticate in any requested tenant; can also list specific tenants here
	AdditionallyAllowedTenants: []string{"*"},
})
```
`GetToken()` returns an error when its arguments specify a tenant the credential isn't configured to allow.

### CODE REVIEW

File: sdk/azidentity/go.mod [hunk 1]
```diff
@@ -2,10 +2,12 @@ module github.com/Azure/azure-sdk-for-go/sdk/azidentity
 
 go 1.18
 
+replace github.com/Azure/azure-sdk-for-go/sdk/azcore => ../azcore
```
Comment (jhendrixMSFT) [inline sdk/azidentity/go.mod:5]: This change is dependent on some unreleased `azcore` functionality?

Comment (chlowell) [reply sdk/azidentity/go.mod:5]: Yes, `TokenRequestOptions.TenantID`

File: sdk/azidentity/azidentity.go [hunk 2]
```diff
@@ -73,6 +75,41 @@ var getPublicClient = func(clientID, tenantID string, co *azcore.ClientOptions)
...
+	return cp
+}
+
+// resolveTenant returns the correct tenant for a token request given a credential's configuration
+func resolveTenant(defaultTenant, reqTenant string, allowedTenants []string) (string, error) {
+	if reqTenant == "" || reqTenant == defaultTenant {
+		return defaultTenant, nil
+	}
+	if defaultTenant == "adfs" {
+		return "", errors.New("the authority doesn't support tenants")
```
Comment (jhendrixMSFT) [inline sdk/azidentity/azidentity.go:100]: Should this error message include the term "adfs" someplace?

File: sdk/azidentity/azure_cli_credential.go [hunk 3]
```diff
@@ -58,8 +63,9 @@ func NewAzureCLICredential(options *AzureCLICredentialOptions) (*AzureCLICredent
 	}
 	cp.init()
 	return &AzureCLICredential{
-		tokenProvider: cp.tokenProvider,
-		tenantID:      cp.TenantID,
+		additionallyAllowedTenants: cp.AdditionallyAllowedTenants,
```
Comment (jhendrixMSFT) [inline sdk/azidentity/azure_cli_credential.go:66]: How come this one doesn't get the `resolveAdditionallyAllowedTenants()` treatment?

Comment (chlowell) [reply sdk/azidentity/azure_cli_credential.go]: I can't think of a reason for it not to, so I believe this is an oversight on the part of past me. Thanks for spotting it.

File: sdk/azidentity/azidentity.go (hunk 2 above)
Comment (chlowell) [reply sdk/azidentity/azidentity.go]: Sure, that could help someone.

**Review by jhendrixMSFT — DISMISSED** [review]

**Comment by chlowell** [comment]
I changed base to `main` because azidentity betas are shipping from there and converted it to a draft pending the azcore beta release.

File: sdk/azidentity/environment_credential.go [hunk 4]
```diff
@@ -75,9 +86,21 @@ func NewEnvironmentCredential(options *EnvironmentCredentialOptions) (*Environme
 	if clientID == "" {
 		return nil, errors.New("missing environment variable " + azureClientID)
 	}
+	// tenants set by NewDefaultAzureCredential() override the value of AZURE_ADDITIONALLY_ALLOWED_TENANTS
+	additionalTenants := options.additionallyAllowedTenants
+	if len(additionalTenants) == 0 {
```
Comment (jhendrixMSFT) [inline sdk/azidentity/environment_credential.go:91]: Nit: Won't this always be true, given that the option isn't exported?

**Review by jhendrixMSFT — APPROVED** [review]

File: sdk/azidentity/environment_credential.go (hunk 4 above)
Comment (chlowell) [reply sdk/azidentity/environment_credential.go:91]: The option exists so `DefaultAzureCredentialOptions.AdditionallyAllowedTenants` can override the environment variable, preventing weirdness like the option having unexpected or inconsistent effects.

**Pull request merged**